### PR TITLE
[CPU] optimize Lp norm for 1-dimensional vector

### LINKED ISSUE #120229: Vector norm significantly slower on 1D data for CPU code
### 🐛 Describe the bug

I was running some math code in PyTorch and noticed that the linalg.norm/norm code was the bottleneck of the performance. Turns out, for one dimensional data, e.g., of shape [n,1], norm is about 50 times slower than with a manual computation. This issue does not occur in higher dimensions, e.g., [n,2], and is easily replicable within colab. This issue only occurs for CPU code, not CUDA code, and regardless of datatype.

```py 
import torch
print('Running 1D:')
# 1D
x = torch.randn(2**18, 1)
y = torch.randn(2**18, 1)
%timeit torch.sqrt((x - y).pow(2).sum(1))
%timeit torch.norm(x - y, 2, 1)
x = x.cuda()
y = y.cuda()
%timeit torch.sqrt((x - y).pow(2).sum(1))
%timeit torch.norm(x - y, 2, 1)

print('Running 2D:')
# 2D
x = torch.randn(2**18, 2)
y = torch.randn(2**18, 2)
%timeit torch.sqrt((x - y).pow(2).sum(1))
%timeit torch.norm(x - y, 2, 1)
x = x.cuda()
y = y.cuda()
%timeit torch.sqrt((x - y).pow(2).sum(1))
%timeit torch.norm(x - y, 2, 1)```
```

Output:
```Running 1D:
1.28 ms ± 151 µs per loop (mean ± std. dev. of 7 runs, 1000 loops each)
33.8 ms ± 727 µs per loop (mean ± std. dev. of 7 runs, 10 loops each)
87.2 µs ± 67 ns per loop (mean ± std. dev. of 7 runs, 10000 loops each)
67.5 µs ± 122 ns per loop (mean ± std. dev. of 7 runs, 10000 loops each)
Running 2D:
3.43 ms ± 453 µs per loop (mean ± std. dev. of 7 runs, 100 loops each)
3.11 ms ± 879 µs per loop (mean ± std. dev. of 7 runs, 100 loops each)
59.9 µs ± 262 ns per loop (mean ± std. dev. of 7 runs, 10000 loops each)
47.9 µs ± 104 ns per loop (mean ± std. dev. of 7 runs, 10000 loops each)```

### Versions

Collecting environment information...
PyTorch version: 2.1.0+cu121
Is debug build: False
CUDA used to build PyTorch: 12.1
ROCM used to build PyTorch: N/A

OS: Ubuntu 22.04.3 LTS (x86_64)
GCC version: (Ubuntu 11.4.0-1ubuntu1~22.04) 11.4.0
Clang version: 14.0.0-1ubuntu1.1
CMake version: version 3.27.9
Libc version: glibc-2.35

Python version: 3.10.12 (main, Nov 20 2023, 15:14:05) [GCC 11.4.0] (64-bit runtime)
Python platform: Linux-6.1.58+-x86_64-with-glibc2.35
Is CUDA available: True
CUDA runtime version: 12.2.140
CUDA_MODULE_LOADING set to: LAZY
GPU models and configuration: GPU 0: Tesla T4
Nvidia driver version: 535.104.05
cuDNN version: Probably one of the following:
/usr/lib/x86_64-linux-gnu/libcudnn.so.8.9.6
/usr/lib/x86_64-linux-gnu/libcudnn_adv_infer.so.8.9.6
/usr/lib/x86_64-linux-gnu/libcudnn_adv_train.so.8.9.6
/usr/lib/x86_64-linux-gnu/libcudnn_cnn_infer.so.8.9.6
/usr/lib/x86_64-linux-gnu/libcudnn_cnn_train.so.8.9.6
/usr/lib/x86_64-linux-gnu/libcudnn_ops_infer.so.8.9.6
/usr/lib/x86_64-linux-gnu/libcudnn_ops_train.so.8.9.6
HIP runtime version: N/A
MIOpen runtime version: N/A
Is XNNPACK available: True

CPU:
Architecture:                       x86_64
CPU op-mode(s):                     32-bit, 64-bit
Address sizes:                      46 bits physical, 48 bits virtual
Byte Order:                         Little Endian
CPU(s):                             2
On-line CPU(s) list:                0,1
Vendor ID:                          GenuineIntel
Model name:                         Intel(R) Xeon(R) CPU @ 2.30GHz
CPU family:                         6
Model:                              63
Thread(s) per core:                 2
Core(s) per socket:                 1
Socket(s):                          1
Stepping:                           0
BogoMIPS:                           4599.99
Flags:                              fpu vme de pse tsc msr pae mce cx8 apic sep mtrr pge mca cmov pat pse36 clflush mmx fxsr sse sse2 ss ht syscall nx pdpe1gb rdtscp lm constant_tsc rep_good nopl xtopology nonstop_tsc cpuid tsc_known_freq pni pclmulqdq ssse3 fma cx16 pcid sse4_1 sse4_2 x2apic movbe popcnt aes xsave avx f16c rdrand hypervisor lahf_lm abm invpcid_single ssbd ibrs ibpb stibp fsgsbase tsc_adjust bmi1 avx2 smep bmi2 erms invpcid xsaveopt arat md_clear arch_capabilities
Hypervisor vendor:                  KVM
Virtualization type:                full
L1d cache:                          32 KiB (1 instance)
L1i cache:                          32 KiB (1 instance)
L2 cache:                           256 KiB (1 instance)
L3 cache:                           45 MiB (1 instance)
NUMA node(s):                       1
NUMA node0 CPU(s):                  0,1
Vulnerability Gather data sampling: Not affected
Vulnerability Itlb multihit:        Not affected
Vulnerability L1tf:                 Mitigation; PTE Inversion
Vulnerability Mds:                  Vulnerable; SMT Host state unknown
Vulnerability Meltdown:             Vulnerable
Vulnerability Mmio stale data:      Vulnerable
Vulnerability Retbleed:             Vulnerable
Vulnerability Spec rstack overflow: Not affected
Vulnerability Spec store bypass:    Vulnerable
Vulnerability Spectre v1:           Vulnerable: __user pointer sanitization and usercopy barriers only; no swapgs barriers
Vulnerability Spectre v2:           Vulnerable, IBPB: disabled, STIBP: disabled, PBRSB-eIBRS: Not affected
Vulnerability Srbds:                Not affected
Vulnerability Tsx async abort:      Not affected

Versions of relevant libraries:
[pip3] numpy==1.25.2
[pip3] torch==2.1.0+cu121
[pip3] torchaudio==2.1.0+cu121
[pip3] torchdata==0.7.0
[pip3] torchsummary==1.5.1
[pip3] torchtext==0.16.0
[pip3] torchvision==0.16.0+cu121
[pip3] triton==2.1.0
[conda] Could not collect

cc @jgong5 @mingfeima @XiaobingSuper @sanchitintel @ashokei @jingxu10 @jianyuh @nikitaved @pearu @mruberry @walterddr @xwang233 @Lezcano

### PR DESCRIPTION
**Pull request by min-jean-cho**
Fixes https://github.com/pytorch/pytorch/issues/120229

- Optimize vector norm by simplifying vector norm formula for 1-dimensional vector. 
- Vector norm formula for 1-dimensional vector simplifies to `abs(x)`. See below for proof. 
- Next step, we can similarly optimize matrix norm (`torch.linalg.matrix_norm`) for 1 x 1 matrix.
- Additionally, avoids overflow in power, `abs(x) ** p` for large `p` or `x`, for 1-dimensional vector.

### Performance
Avg Latency (ms) of `torch.norm` and `torch.linalg.vector_norm` for
`torch.norm(torch.randn(2**18, 1), ord, -1)`
`torch.linalg.vector_norm(torch.randn(2**18, 1), ord, -1)`
Tested on 28 physical cores/socket, 1 socket on Skylake.

|                          	|                 	|         	|         	| **Avg Latency (ms)**  	|                       	|                                        	|
|--------------------------	|-----------------	|---------	|---------	|-----------------------	|-----------------------	|----------------------------------------	|
| **op**                   	| **input shape** 	| **dim** 	| **ord** 	| **baseline (master)** 	| **optimized (7102f1ef372b248414d36cbd0c51a546b6b6a41a)** 	| **speedup ratio (baseline/optimized)** 	|
| torch.norm               	| (2**18, 1)      	| -1      	| fro     	| 34.3755531            	| 0.0125408             	| 2741.094                               	|
|                          	|                 	|         	| inf     	| 34.0952635            	| 0.0122237             	| 2789.271                               	|
|                          	|                 	|         	| -inf    	| 34.3674493            	| 0.0120759             	| 2845.953                               	|
|                          	|                 	|         	| 0       	| 34.1004515            	| 0.0175261             	| 1945.69                                	|
|                          	|                 	|         	| 1       	| 34.1688442            	| 0.0121593             	| 2810.089                               	|
|                          	|                 	|         	| -1      	| 33.949492             	| 0.0120282             	| 2822.487                               	|
|                          	|                 	|         	| 2       	| 34.3669581            	| 0.0120401             	| 2854.366                               	|
|                          	|                 	|         	| -2      	| 33.9252067            	| 0.0121069             	| 2802.139                               	|
|                          	|                 	|         	|         	|                       	|                       	|                                        	|
| torch.linalg.vector_norm 	| (2**18, 1)      	| -1      	| inf     	| 34.090879             	| 0.0095105             	| 3584.545                               	|
|                          	|                 	|         	| -inf    	| 34.3708754            	| 0.0099111             	| 3467.931                               	|
|                          	|                 	|         	| 0       	| 34.0880775            	| 0.0141716             	| 2405.38                                	|
|                          	|                 	|         	| 1       	| 34.1392851            	| 0.0093174             	| 3664.036                               	|
|                          	|                 	|         	| -1      	| 33.925395             	| 0.0092483             	| 3668.302                               	|
|                          	|                 	|         	| 2       	| 34.3854165            	| 0.0092459             	| 3719.002                               	|
|                          	|                 	|         	| -2      	| 33.932972             	| 0.0093007             	| 3648.429                               	|

### Proof
<details>
<summary>For those interested :)</summary>

1_dim_vector_norm_proof1
1_dim_vector_norm_proof2

</details>

cc @jgong5 @mingfeima @XiaobingSuper @sanchitintel @ashokei @jingxu10 @jianyuh @nikitaved @pearu @mruberry @walterddr @xwang233 @Lezcano @voznesenskym @penguinwu @EikanWang @Guobing-Chen @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @peterbell10 @ipiszy @yf225 @chenyang78 @kadeng @muchulee8 @aakhundov @ColinPeppler @amjames @desertfire @chauhang

### CODE REVIEW

File: test/test_foreach.py [hunk 1]
```diff
@@ -631,7 +631,7 @@ def test_foreach_l2_large_value_input(self, device, dtype, op):
         scaler = torch.tensor([max_value]).sqrt().to(device=device, dtype=dtype)
         inputs = ([
             t * scaler for t in next(iter(op.sample_inputs(device, dtype, requries_grad=True, num_input_tensors=[N], low=1))).input
-        ],)
+        ][:-1],)
```
Comment (min-jean-cho) [inline test/test_foreach.py:634]: test was failing as the vector norm reformulation prevents overflow in power, `abs(x_i) ** p` for large `p` or `x_i`, for 1-dimensional vector https://github.com/pytorch/pytorch/blob/139647d317ace829a0993b459411fa59191cb617/test/test_foreach.py#L645-L647

```python
import torch
import numpy as np

x = torch.tensor([[1.90232e+19]], dtype=torch.float32)
x_np = np.array([[1.90232e+19]], dtype=np.float32)

norm = torch.linalg.vector_norm(x, ord=2)
norm_np = np.linalg.norm(x_np, ord=2)

print(norm) # before: tensor(inf)   after: tensor(1.9023e+19) 
print(norm_np) # 1.90232e+19
```

File: aten/src/ATen/native/LinearAlgebra.cpp [hunk 2]
```diff
@@ -2815,6 +2815,61 @@ TORCH_IMPL_FUNC(linalg_vector_norm_out)(const Tensor& self, const Scalar& scalar
   // values larger than 10^53 (same for negative numbers), so that's fine.
   auto ord = scalar_ord.toDouble();
   auto dim = opt_dim.value_or(IntArrayRef{});
+  auto size = self.sizes();
+  auto ndim = self.dim();
+
+  bool all_reduction_dims_are_one_dimensional = true;
+  if (result.sizes().empty()){
```
Comment (lezcano) [inline aten/src/ATen/native/LinearAlgebra.cpp:2822]: No need for this path really

File: aten/src/ATen/native/LinearAlgebra.cpp [hunk 3]
```diff
@@ -2815,6 +2815,61 @@ TORCH_IMPL_FUNC(linalg_vector_norm_out)(const Tensor& self, const Scalar& scalar
...
+      DimVector dims_ = at::native::make_dim_vector(opt_dim, self.dim());
+      maybe_wrap_dims(dims_, self.dim());
+      for (const auto i : c10::irange(dims_.size())) {
+        size_t dim_ = dims_[i];
+        if (size[dim_] != 1){
+          all_reduction_dims_are_one_dimensional = false;
+          break;
+        }
+      }
+    } else {
```
Comment (lezcano) [inline aten/src/ATen/native/LinearAlgebra.cpp:2843]: Pretty sure this is dead code, as if the result is not 0-dim, it must have `dim` specified.

File: aten/src/ATen/native/LinearAlgebra.cpp [hunk 4]
```diff
@@ -2815,6 +2815,61 @@ TORCH_IMPL_FUNC(linalg_vector_norm_out)(const Tensor& self, const Scalar& scalar
...
+      all_reduction_dims_are_one_dimensional = false;
+    }
+  }
+
+  if (all_reduction_dims_are_one_dimensional && !self.is_complex()) {
+    Tensor result_tmp = self.clone();
```
Comment (lezcano) [inline aten/src/ATen/native/LinearAlgebra.cpp:2849]: In most cases you are doing a clone + abs + copy. You can do just one operation using `abs_out`. Similar ofr the `ord == 0` case. Just do `at::abs_outf(self.squeeze(dim), result)`

**Review by lezcano — COMMENTED** [review]
Fair, but the implementation can be improved and simplified.

Can you also add a test?

**Review by lezcano — COMMENTED** [review]
Also, now that you are at it, can you update the decomposition as well?

**Comment by min-jean-cho** [comment]
Thank you @lezcano, updated with cleaner implementation.

File: aten/src/ATen/native/LinearAlgebra.cpp (hunk 4 above)
Comment (min-jean-cho) [reply aten/src/ATen/native/LinearAlgebra.cpp]: Thank you @lezcano, updated performance data as well.

File: aten/src/ATen/native/LinearAlgebra.cpp [hunk 5]
```diff
@@ -2815,6 +2817,36 @@ TORCH_IMPL_FUNC(linalg_vector_norm_out)(const Tensor& self, const Scalar& scalar
...
+
+  bool is_reduce_over_1D_vector = true;
+  for (auto i : reduce_dim) {
+    if (size[i] != 1){
+      is_reduce_over_1D_vector = false;
+      break;
+    }
+  }
+
+  if (is_reduce_over_1D_vector && !self.is_complex()) {
```
Comment (lezcano) [inline aten/src/ATen/native/LinearAlgebra.cpp:2841]: This also works for complex numbers.

**Review by lezcano — COMMENTED** [review]
This looks much better. Can you now add a test and update the decomposition?

ps. I see you provide a case-by-case proof of this fact. A much simpler way of proving this is realising that a norm on R is given by its value at 1 by the homogeneity properties of norms. As such, any norm on the real numbers is of the form |c| * ||1||.  You just need to see that all these norms map 1 to 1 and you are done. While doing this, you also find that the "0 norm" is not actually a norm.

File: aten/src/ATen/native/LinearAlgebra.cpp (hunk 5 above)
Comment (min-jean-cho) [reply aten/src/ATen/native/LinearAlgebra.cpp]: Thank you @lezcano, you're right.

**Comment by min-jean-cho** [comment]
Thank you @lezcano. May I follow-up with the decomposition in a separate PR? 

ps. That's nice, thanks for the background :)

**Comment by lezcano** [comment]
I think updating the decomp shouldn't take longer than two lines. You can even golf it as
```python
is_reduce_over_1D_vector = (dim is None and x.numel() == 1) or all(shape[i] == 1 for i in dim)
```
This trick tan also be used in the C++ impl, but the C++ version will not read as nicely as the Python one :)

File: test/test_linalg.py [hunk 6]
```diff
@@ -1631,6 +1631,29 @@ def test_vector_norm_extreme_values(self, device):
...
+            ((1,), None),
+        ]
+        orders = [float('inf'), -float('inf'), 0, 1, -1, 2, -2]
+        keepdims = [True, False]
+
+        for input_size_and_dim, ord, keepdim in product(input_sizes_and_dims, orders, keepdims):
+            input_size = input_size_and_dim[0]
+            dim = input_size_and_dim[1]
+            if type(dim) is tuple and ord == 0:
+                # skip because np.linalg.norm raises 'ValueError: Invalid norm order for matrices.'
```
Comment (lezcano) [inline test/test_linalg.py:1648]: cc @rgommers

**Review by lezcano — APPROVED** [review]
SG. Please tag me in the follow-up PR updating the decomposition

**Comment by min-jean-cho** [comment]
Thank you @lezcano for the review and feedbacks. I will tag you.

**Pull request closed without merging**